File: division-id-tool.php
<?php
// ==============================
// division_id_tool.php — Version PHP autonome
// Aucune dépendance externe requise - utilise uniquement les fonctions standard PHP
// ==============================

// Préfixes typologiques francophones
$PREFIX_MAP = [
    "pays" => "PA", "region" => "RG", "prefecture" => "PR", "province" => "PV",
    "departement" => "DP", "district" => "DI", "arrondissement" => "AR", "commune" => "CM",
    "ville" => "VL", "quartier" => "QR", "localite" => "LC", "secteur-villageois" => "SV",
    "village" => "VG", "zone-sanitaire" => "ZS", "zone-de-developpement" => "ZD",
    "zone-electorale" => "ZE", "canton" => "CC", "territoire" => "TR", "metropole" => "MT"
];

// Custom slugify function to match JavaScript and Python behavior
function slugify($text, $separator = "-", $lowercase = true) {
    // Normalize Unicode characters (decompose)
    if (class_exists('Normalizer')) {
        $text = Normalizer::normalize($text, Normalizer::FORM_D);
    }
    
    // Remove accents (combining characters)
    $text = preg_replace('/[\x{0300}-\x{036F}]/u', '', $text);
    
    if ($lowercase) {
        $text = mb_strtolower($text, 'UTF-8');
    }
    
    // Replace non-alphanumeric characters with separator
    $text = preg_replace('/[^a-zA-Z0-9]+/', $separator, $text);
    
    // Remove leading/trailing separators
    $text = trim($text, $separator);
    
    return $text;
}

// RFC4648 Base32 encoding function
function rfc4648Base32Encode($data) {
    $alphabet = 'ABCDEFGHIJKLMNOPQRSTUVWXYZ234567';
    $result = '';
    $bits = 0;
    $value = 0;
    
    for ($i = 0; $i < strlen($data); $i++) {
        $value = ($value << 8) | ord($data[$i]);
        $bits += 8;
        
        while ($bits >= 5) {
            $result .= $alphabet[($value >> ($bits - 5)) & 31];
            $bits -= 5;
        }
    }
    
    if ($bits > 0) {
        $result .= $alphabet[($value << (5 - $bits)) & 31];
    }
    
    return $result;
}

function normalizedSlug($pathArray) {
    $slugs = [];
    foreach ($pathArray as $name) {
        $slugs[] = slugify($name, "-", true);
    }
    return implode(".", $slugs);
}

function generateDivisionId($prefix, $pathArray, $prefixDelimiter = '') {
    $slug = normalizedSlug($pathArray);
    $hash = hash('sha256', $slug, true);
    $base32Id = rfc4648Base32Encode($hash);
    
    return [
        'division_id' => $prefix . $prefixDelimiter . substr($base32Id, 0, 6),
        'division_slug' => $slug
    ];
}

function generateSqlInserts($inputFile, $outputFile) {
    global $PREFIX_MAP;
    
    if (!file_exists($inputFile)) {
        echo "❌ Fichier d'entrée non trouvé: $inputFile\n";
        exit(1);
    }
    
    // Read CSV headers
    $handle = fopen($inputFile, 'r');
    if ($handle === false) {
        echo "❌ Impossible d'ouvrir le fichier: $inputFile\n";
        exit(1);
    }
    
    $headers = fgetcsv($handle, 0, ',', '"', '\\');
    if ($headers === false) {
        echo "❌ Impossible de lire les en-têtes du fichier CSV\n";
        fclose($handle);
        exit(1);
    }
    
    // Generate division types from headers
    $divisionTypes = [];
    $divisionTypeMap = [];
    $typeIdCounter = 1;
    
    foreach ($headers as $level => $headerName) {
        $typeSlug = slugify($headerName, "-", true);
        
        if (!isset($PREFIX_MAP[$typeSlug])) {
            echo "❌ Type de division inconnu: '$typeSlug'. Veuillez mettre à jour le tableau de correspondance.\n";
            exit(1);
        }
        
        $prefix = $PREFIX_MAP[$typeSlug];
        $parentTypeId = $level > 0 ? $level : 'null';
        
        $typeInfo = [
            'id' => $typeIdCounter,
            'name' => $headerName,
            'code' => $prefix,
            'level' => $level,
            'parent_type_id' => $parentTypeId,
            'description' => "Type de division: $headerName"
        ];
        
        $divisionTypes[] = $typeInfo;
        $divisionTypeMap[$headerName] = $typeInfo;
        $typeIdCounter++;
    }
    
    // Generate divisions with parent relationships
    $divisions = [];
    $divisionMap = [];
    
    // Reset to beginning
    rewind($handle);
    fgetcsv($handle, 0, ',', '"', '\\'); // Skip header
    
    while (($row = fgetcsv($handle, 0, ',', '"', '\\')) !== false) {
        // Ensure we have values for all headers, padding with empty strings if needed
        $fullRow = [];
        for ($i = 0; $i < count($headers); $i++) {
            $fullRow[] = isset($row[$i]) ? trim($row[$i]) : '';
        }
        
        $pathArray = array_filter($fullRow, function($v) { return $v !== ''; });
        
        // Skip if no data
        if (count($pathArray) === 0) continue;
        
        // Validate: no empty columns in the middle
        $lastIdx = -1;
        for ($i = 1; $i < count($fullRow); $i++) {
            if ($fullRow[$i] === '' && $fullRow[$i-1] !== '') {
                $lastIdx = $i;
                break;
            }
        }
        
        $isValid = $lastIdx === -1 || array_reduce(
            array_slice($fullRow, $lastIdx), 
            function($carry, $item) { return $carry && $item === ''; }, 
            true
        );
        
        if (!$isValid) continue;
        
        $pathArray = array_values($pathArray); // Re-index array
        $typeName = $headers[count($pathArray) - 1];
        $typeSlug = slugify($typeName, "-", true);
        $prefix = $PREFIX_MAP[$typeSlug];
        
        $divisionData = generateDivisionId($prefix, $pathArray, '-');
        
        // Calculate parent_id by getting the division_id of the first n-1 elements
        $parentId = null;
        if (count($pathArray) > 1) {
            $parentPathArray = array_slice($pathArray, 0, -1);
            $parentTypeName = $headers[count($parentPathArray) - 1];
            $parentTypeSlug = slugify($parentTypeName, "-", true);
            $parentPrefix = $PREFIX_MAP[$parentTypeSlug];
            $parentDivisionData = generateDivisionId($parentPrefix, $parentPathArray, '-');
            $parentId = $parentDivisionData['division_id'];
        }
        
        $typeInfo = $divisionTypeMap[$typeName];
        
        $division = [
            'id' => $divisionData['division_id'],
            'type_id' => $typeInfo['id'],
            'parent_id' => $parentId,
            'name' => $pathArray[count($pathArray) - 1], // Last element is the name
            'division_slug' => $divisionData['division_slug']
        ];
        
        $divisions[] = $division;
        $divisionMap[$divisionData['division_id']] = $division;
    }
    
    fclose($handle);
    
    // Generate SQL
    $sql = "-- SQL Insert Script for Country Tree Structure\n";
    $sql .= "-- Generated by division-id-tool.php\n\n";
    
    // Insert division types
    $sql .= "-- Insert division types\n";
    foreach ($divisionTypes as $type) {
        $name = str_replace("'", "''", $type['name']);
        $description = str_replace("'", "''", $type['description']);
        $sql .= "INSERT INTO country_division_types (id, name, code, level, parent_type_id, description) VALUES ";
        $sql .= "({$type['id']}, '{$name}', '{$type['code']}', {$type['level']}, {$type['parent_type_id']}, '{$description}');\n";
    }
    
    $sql .= "\n-- Insert divisions\n";
    foreach ($divisions as $division) {
        $parentIdValue = $division['parent_id'] ? "'{$division['parent_id']}'" : 'NULL';
        $name = str_replace("'", "''", $division['name']);
        $sql .= "INSERT INTO country_divisions (id, type_id, parent_id, name, division_slug) VALUES ";
        $sql .= "('{$division['id']}', {$division['type_id']}, {$parentIdValue}, '{$name}', '{$division['division_slug']}');\n";
    }
    
    // Write SQL to file
    if (file_put_contents($outputFile, $sql) === false) {
        echo "❌ Impossible de créer le fichier de sortie: $outputFile\n";
        exit(1);
    }
    
    echo "✅ Script SQL généré : $outputFile\n";
}

function processCsv($inputFile, $outputFile) {
    global $PREFIX_MAP;
    
    if (!file_exists($inputFile)) {
        echo "❌ Fichier d'entrée non trouvé: $inputFile\n";
        exit(1);
    }
    
    $results = [];
    $headers = [];
    
    // Read and parse CSV
    $handle = fopen($inputFile, 'r');
    if ($handle === false) {
        echo "❌ Impossible d'ouvrir le fichier: $inputFile\n";
        exit(1);
    }
    
    // Read headers
    $headers = fgetcsv($handle, 0, ',', '"', '\\');
    if ($headers === false) {
        echo "❌ Impossible de lire les en-têtes du fichier CSV\n";
        fclose($handle);
        exit(1);
    }
    
    // Process each row
    while (($row = fgetcsv($handle, 0, ',', '"', '\\')) !== false) {
        // Ensure we have values for all headers, padding with empty strings if needed
        $fullRow = [];
        for ($i = 0; $i < count($headers); $i++) {
            $fullRow[] = isset($row[$i]) ? trim($row[$i]) : '';
        }
        
        $pathArray = array_filter($fullRow, function($v) { return $v !== ''; });
        
        // Skip if no data
        if (count($pathArray) === 0) continue;
        
        // Validate: no empty columns in the middle
        $lastIdx = -1;
        for ($i = 1; $i < count($fullRow); $i++) {
            if ($fullRow[$i] === '' && $fullRow[$i-1] !== '') {
                $lastIdx = $i;
                break;
            }
        }
        
        $isValid = $lastIdx === -1 || array_reduce(
            array_slice($fullRow, $lastIdx), 
            function($carry, $item) { return $carry && $item === ''; }, 
            true
        );
        
        if (!$isValid) continue;
        
        $typeName = $headers[count($pathArray) - 1];
        $typeSlug = slugify($typeName, "-", true);
        
        if (!isset($PREFIX_MAP[$typeSlug])) {
            echo "❌ Type de division inconnu: '$typeSlug'. Veuillez mettre à jour le tableau de correspondance.\n";
            exit(1);
        }
        
        $prefix = $PREFIX_MAP[$typeSlug];
        $divisionData = generateDivisionId($prefix, $pathArray, '-');
        
        // Calculate parent_id
        $parentId = null;
        if (count($pathArray) > 1) {
            $parentPathArray = array_slice($pathArray, 0, -1);
            $parentTypeName = $headers[count($parentPathArray) - 1];
            $parentTypeSlug = slugify($parentTypeName, "-", true);
            $parentPrefix = $PREFIX_MAP[$parentTypeSlug];
            $parentDivisionData = generateDivisionId($parentPrefix, $parentPathArray, '-');
            $parentId = $parentDivisionData['division_id'];
        }
        
        // Create enriched row
        $enriched = array_combine($headers, $fullRow);
        $enriched['parent_id'] = $parentId;
        $enriched['division_id'] = $divisionData['division_id'];
        $enriched['division_slug'] = $divisionData['division_slug'];
        
        $results[] = $enriched;
    }
    
    fclose($handle);
    
    if (count($results) === 0) {
        echo "⚠️ Aucun enregistrement valide trouvé.\n";
        return;
    }
    
    // Determine the maximum depth used in the data
    $maxDepth = 0;
    foreach ($results as $row) {
        $fullRow = [];
        foreach ($headers as $header) {
            $fullRow[] = trim($row[$header] ?? '');
        }
        $pathArray = array_filter($fullRow, function($v) { return $v !== ''; });
        $maxDepth = max($maxDepth, count($pathArray));
    }
    
    // Only include headers up to the maximum depth used + division fields
    $usedHeaders = array_slice($headers, 0, $maxDepth);
    $extendedHeaders = array_merge($usedHeaders, ['parent_id', 'division_id', 'division_slug']);
    
    // Write CSV output
    $outputHandle = fopen($outputFile, 'w');
    if ($outputHandle === false) {
        echo "❌ Impossible de créer le fichier de sortie: $outputFile\n";
        exit(1);
    }
    
    // Write headers
    fputcsv($outputHandle, $extendedHeaders, ',', '"', '\\');
    
    // Write data rows
    foreach ($results as $row) {
        $outputRow = [];
        foreach ($extendedHeaders as $header) {
            $outputRow[] = $row[$header] ?? '';
        }
        fputcsv($outputHandle, $outputRow, ',', '"', '\\');
    }
    
    fclose($outputHandle);
    echo "✅ Fichier généré : $outputFile\n";
}

// Exécution directe
if (php_sapi_name() === 'cli') {
    // Check for SQL mode
    if (in_array('--sql', $argv)) {
        $sqlIndex = array_search('--sql', $argv);
        $inputFile = $argv[$sqlIndex + 1] ?? "data.csv";
        $outputFile = $argv[$sqlIndex + 2] ?? "inserts.sql";
        
        generateSqlInserts($inputFile, $outputFile);
    } else {
        // CSV mode
        $inputFile = $argv[1] ?? "data.csv";
        $outputFile = $argv[2] ?? "output.csv";
        
        processCsv($inputFile, $outputFile);
    }
}

File: division-id-tool.py
# ==============================
# division_id_tool.py — Version Python adaptée multi-niveaux
# Aucune dépendance externe requise - utilise uniquement les modules standard Python
# ==============================

import hashlib
import csv
import base64
import sys
import re
import unicodedata

# Préfixes typologiques francophones
PREFIX_MAP = {
    "pays": "PA", "region": "RG", "prefecture": "PR", "province": "PV",
    "departement": "DP", "district": "DI", "arrondissement": "AR", "commune": "CM",
    "ville": "VL", "quartier": "QR", "localite": "LC", "secteur-villageois": "SV",
    "village": "VG", "zone-sanitaire": "ZS", "zone-de-developpement": "ZD",
    "zone-electorale": "ZE", "canton": "CC", "territoire": "TR", "metropole": "MT"
}

def slugify(text, separator="-", lowercase=True):
    """Simple slugify function to match JavaScript slugify behavior"""
    # Convert to NFD (decomposed) form and remove accents
    text = unicodedata.normalize('NFD', text)
    text = ''.join(char for char in text if unicodedata.category(char) != 'Mn')
    
    if lowercase:
        text = text.lower()
    
    # Replace non-alphanumeric characters with separator
    text = re.sub(r'[^a-zA-Z0-9]+', separator, text)
    
    # Remove leading/trailing separators
    text = text.strip(separator)
    
    return text

def normalized_slug(path):
    return ".".join([slugify(name, separator="-", lowercase=True) for name in path])

def generate_division_id(prefix, path, prefix_delimiter=''):
    slug = normalized_slug(path)
    digest = hashlib.sha256(slug.encode()).digest()
    base32_id = base64.b32encode(digest).decode("utf-8").upper().rstrip("=")
    return f"{prefix}{prefix_delimiter}{base32_id[:6]}", slug

def generate_sql_inserts(input_file, output_file):
    """Generate SQL insert statements compatible with the database schema"""
    with open(input_file, 'r', encoding='utf-8') as csvfile:
        reader = csv.DictReader(csvfile)
        headers = reader.fieldnames
        
        # Generate division types from headers
        division_types = []
        division_type_map = {}
        type_id_counter = 1
        
        for level, header_name in enumerate(headers):
            type_slug = slugify(header_name, separator="-", lowercase=True)
            prefix = PREFIX_MAP.get(type_slug)
            
            if not prefix:
                print(f"❌ Type de division inconnu: '{type_slug}'. Veuillez mettre à jour le tableau de correspondance.")
                sys.exit(1)
            
            parent_type_id = level if level > 0 else 'null'
            
            type_info = {
                'id': type_id_counter,
                'name': header_name,
                'code': prefix,
                'level': level,
                'parent_type_id': parent_type_id,
                'description': f"Type de division: {header_name}"
            }
            
            division_types.append(type_info)
            division_type_map[header_name] = type_info
            type_id_counter += 1
        
        # Generate divisions with parent relationships
        divisions = []
        division_map = {}
        
        csvfile.seek(0)  # Reset to beginning
        next(reader)  # Skip header
        
        for row in reader:
            full_row = [str(row.get(col, "") or "").strip() for col in headers]
            path_array = [v for v in full_row if v != ""]
            
            # Skip if no data
            if len(path_array) == 0:
                continue
            
            # Validate: no empty columns in the middle
            last_idx = -1
            for i in range(1, len(full_row)):
                if full_row[i] == "" and full_row[i-1] != "":
                    last_idx = i
                    break
            
            is_valid = last_idx == -1 or all(v == "" for v in full_row[last_idx:])
            if not is_valid:
                continue
            
            type_name = headers[len(path_array)-1]
            type_slug = slugify(type_name, separator="-", lowercase=True)
            prefix = PREFIX_MAP.get(type_slug)
            
            division_id, division_slug = generate_division_id(prefix, path_array, '-')
            
            # Calculate parent_id by getting the division_id of the first n-1 elements
            parent_id = None
            if len(path_array) > 1:
                parent_path_array = path_array[:-1]
                parent_type_name = headers[len(parent_path_array)-1]
                parent_type_slug = slugify(parent_type_name, separator="-", lowercase=True)
                parent_prefix = PREFIX_MAP.get(parent_type_slug)
                parent_division_id, _ = generate_division_id(parent_prefix, parent_path_array, '-')
                parent_id = parent_division_id
            
            type_info = division_type_map[type_name]
            
            division = {
                'id': division_id,
                'type_id': type_info['id'],
                'parent_id': parent_id,
                'name': path_array[-1],  # Last element is the name
                'division_slug': division_slug
            }
            
            divisions.append(division)
            division_map[division_id] = division
    
    # Generate SQL
    sql = "-- SQL Insert Script for Country Tree Structure\n"
    sql += "-- Generated by division-id-tool.py\n\n"
    
    # Insert division types
    sql += "-- Insert division types\n"
    for type_info in division_types:
        name_escaped = type_info['name'].replace("'", "''")
        description_escaped = type_info['description'].replace("'", "''")
        sql += f"INSERT INTO country_division_types (id, name, code, level, parent_type_id, description) VALUES "
        sql += f"({type_info['id']}, '{name_escaped}', '{type_info['code']}', "
        sql += f"{type_info['level']}, {type_info['parent_type_id']}, '{description_escaped}');\n"
    
    sql += "\n-- Insert divisions\n"
    for division in divisions:
        parent_id_value = f"'{division['parent_id']}'" if division['parent_id'] else 'NULL'
        name_escaped = division['name'].replace("'", "''")
        sql += f"INSERT INTO country_divisions (id, type_id, parent_id, name, division_slug) VALUES "
        sql += f"('{division['id']}', {division['type_id']}, {parent_id_value}, "
        sql += f"'{name_escaped}', '{division['division_slug']}');\n"
    
    # Write SQL to file
    with open(output_file, 'w', encoding='utf-8') as f:
        f.write(sql)
    
    print(f"✅ Script SQL généré : {output_file}")

def process_csv(input_file, output_file):
    enriched_rows = []
    headers = []
    
    with open(input_file, 'r', encoding='utf-8') as csvfile:
        reader = csv.DictReader(csvfile)
        headers = reader.fieldnames
        
        for row in reader:
            full_row = [str(row.get(col, "") or "").strip() for col in headers]
            path_array = [v for v in full_row if v != ""]
            
            # Skip if no data
            if len(path_array) == 0:
                continue

            # Validate: no empty columns in the middle
            last_idx = -1
            for i in range(1, len(full_row)):
                if full_row[i] == "" and full_row[i-1] != "":
                    last_idx = i
                    break
            
            is_valid = last_idx == -1 or all(v == "" for v in full_row[last_idx:])
            if not is_valid:
                continue
                
            type_name = headers[len(path_array)-1]
            type_slug = slugify(type_name, separator="-", lowercase=True)
            prefix = PREFIX_MAP.get(type_slug)
            if not prefix:
                print(f"❌ Type de division inconnu: '{type_slug}'. Veuillez mettre à jour le tableau de correspondance.")
                sys.exit(1)

            division_id, slug = generate_division_id(prefix, path_array, '-')
            
            # Calculate parent_id
            parent_id = None
            if len(path_array) > 1:
                parent_path_array = path_array[:-1]
                parent_type_name = headers[len(parent_path_array)-1]
                parent_type_slug = slugify(parent_type_name, separator="-", lowercase=True)
                parent_prefix = PREFIX_MAP.get(parent_type_slug)
                parent_division_id, _ = generate_division_id(parent_prefix, parent_path_array, '-')
                parent_id = parent_division_id
            
            enriched = dict(row)
            enriched["parent_id"] = parent_id
            enriched["division_id"] = division_id
            enriched["division_slug"] = slug
            enriched_rows.append(enriched)

    if enriched_rows:
        # Determine the maximum depth used in the data
        max_depth = max(len([v for v in [str(row.get(col, "")).strip() for col in headers] if v != ""]) 
                        for row in enriched_rows)
        
        # Only include headers up to the maximum depth used + division fields
        used_headers = headers[:max_depth]
        extended_headers = used_headers + ['parent_id', 'division_id', 'division_slug']
        
        # Write CSV output
        with open(output_file, 'w', encoding='utf-8', newline='') as csvfile:
            writer = csv.DictWriter(csvfile, fieldnames=extended_headers)
            writer.writeheader()
            writer.writerows(enriched_rows)
            
        print(f"✅ Fichier généré : {output_file}")
    else:
        print("⚠️ Aucun enregistrement valide trouvé.")

if __name__ == "__main__":
    # Check for SQL mode
    if '--sql' in sys.argv:
        sql_index = sys.argv.index('--sql')
        infile = sys.argv[sql_index + 1] if len(sys.argv) > sql_index + 1 else "data.csv"
        outfile = sys.argv[sql_index + 2] if len(sys.argv) > sql_index + 2 else "inserts.sql"
        generate_sql_inserts(infile, outfile)
    else:
        # CSV mode
        infile = sys.argv[1] if len(sys.argv) > 1 else "data.csv"
        outfile = sys.argv[2] if len(sys.argv) > 2 else "output.csv"
        process_csv(infile, outfile)